### ISSUE TITLE
libnvidia-container-1.0.6.drv fails to build: missing rpcgen

### ISSUE THREAD

**Issue by rayhogenson** [NixOS/nixpkgs#101517]
**Describe the bug**
When I tried to run a system upgrade, it failed to build libnvidia-container. Trying to build this derivation separately also fails. I guess a missing dependency.
```
$ nix build /nix/store/gw0dz18pykg8zimmr3hs856szh1xqc92-libnvidia-container-1.0.6.drv
builder for '/nix/store/gw0dz18pykg8zimmr3hs856szh1xqc92-libnvidia-container-1.0.6.drv' failed with exit code 2; last 10 log lines:
  patching file src/nvc_info.c
  patching file src/nvc_ldcache.c
  configuring
  no configure script, doing nothing
  building
  build flags: SHELL=/nix/store/k8p54jg8ipvnfz435mayf5bnqhw4qqap-bash-4.4-p23/bin/bash WITH_LIBELF=yes prefix=\$\(out\)
  rm -f /build/source/src/driver_rpc.h
  cd /build/source/src/ && rpcgen -h -C -M -N -o driver_rpc.h /build/source/src/driver_rpc.x
  /nix/store/k8p54jg8ipvnfz435mayf5bnqhw4qqap-bash-4.4-p23/bin/bash: rpcgen: command not found
  make: *** [Makefile:174: /build/source/src/driver_rpc.h] Error 127
[0 built (1 failed)]
error: build of '/nix/store/gw0dz18pykg8zimmr3hs856szh1xqc92-libnvidia-container-1.0.6.drv' failed
```

**To Reproduce**
1. nix build /nix/store/gw0dz18pykg8zimmr3hs856szh1xqc92-libnvidia-container-1.0.6.drv

**Expected behavior**
libnvidia-container should build without errors.

**Additional context**

**Notify maintainers**

**Metadata**
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/hfvjs6abgzadrv9f387v1ip5mr206jh9-bash-interactive-4.4-p23-dev
copying path '/nix/store/hfvjs6abgzadrv9f387v1ip5mr206jh9-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.70, NixOS, 21.03pre246543.24c9b05ac53 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.7`
 - channels(root): `"nixos-21.03pre248170.007126eef72"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by GuillaumeDesforges** [comment]
Fails to build for me as well

Seems to be missing rpcgen in buildInputs

https://github.com/NixOS/nixpkgs/blob/4a704c653b06b423b771f92af4f5c069df7dd682/pkgs/applications/virtualization/nvidia-docker/libnvc.nix#L45-L47

**Comment by GuillaumeDesforges** [comment]
I can also see the same problem here:

https://github.com/NixOS/nixpkgs/issues/89166

After looking around, I can't seem to find a proper rpcgen. I tried the following:

```nix
{ stdenv, lib, fetchFromGitHub, pkgconfig, libelf, libcap, libseccomp }:

with lib; let

  modp-ver = "396.51";

  nvidia-modprobe = fetchFromGitHub {
    owner = "NVIDIA";
    repo = "nvidia-modprobe";
    rev = modp-ver;
    sha256 = "1fw2qwc84k64agw6fx2v0mjf88aggph9c6qhs4cv7l3gmflv8qbk";
  };

in stdenv.mkDerivation rec {
  pname = "libnvidia-container";
  version = "1.0.6";

  src = fetchFromGitHub {
    owner = "NVIDIA";
    repo = "libnvidia-container";
    rev = "v${version}";
    sha256 = "1pnpc9knwh8d1zqb28zc3spkjc00w0z10vd3jna8ksvpl35jl7w3";
  };

  # locations of nvidia-driver libraries are not resolved via ldconfig which
  # doesn't get used on NixOS. Additional support binaries like nvidia-smi are
  # not resolved via the environment PATH but via the derivation output path.
  patches = [ ./libnvc-ldconfig-and-path-fixes.patch ];

  makeFlags = [
    "WITH_LIBELF=yes"
    "prefix=$(out)"
  ];

  postPatch = ''
    sed -i 's/^REVISION :=.*/REVISION = ${src.rev}/' mk/common.mk
    sed -i 's/^COMPILER :=.*/COMPILER = $(CC)/' mk/common.mk

    mkdir -p deps/src/nvidia-modprobe-${modp-ver}
    cp -r ${nvidia-modprobe}/* deps/src/nvidia-modprobe-${modp-ver}
    chmod -R u+w deps/src
    touch deps/src/nvidia-modprobe-${modp-ver}/.download_stamp
  '';

  nativeBuildInputs = [ pkgconfig ];

  buildInputs = [ libelf libcap libseccomp ];

  meta = {
    homepage = "https://github.com/NVIDIA/libnvidia-container";
    description = "NVIDIA container runtime library";
    license = licenses.bsd3;
    platforms = platforms.linux;
  };
}
```

that provides a rpcgen, but got errors

```
gcc -fPIC -std=gnu11 -O2 -g -fdata-sections -ffunction-sections -fstack-protector -fno-strict-aliasing -fvisibility=hidden -Wall -Wextra -Wcast-align -Wpointer-arith -Wmissing-prototypes -Wnonnull -Wwrite-strings -Wlogical-op -Wformat=2 -Wmissing-format-attribute -Winit-self -Wshadow -Wstrict-prototypes -Wunreachable-code -Wconversion -Wsign-conversion -Wno-unknown-warning-option -Wno-format-extra-args -Wno-gnu-alignof-expression  -DNV_LINUX -isystem /build/source/deps/nix/store/chz7gx3jbzqpsaj5hxyf2a7ws294q33i-libnvidia-container-1.0.6/include -include /build/source/src/build.h -DWITH_LIBELF -DWITH_SECCOMP -I/nix/store/mmcj5dvbjc6d9rac1hb0an658s3lbd0q-libseccomp-2.5.0-dev/include -D_GNU_SOURCE -D_FORTIFY_SOURCE=2  -DNDEBUG -MMD -MF /build/source/src/driver.d -c -o /build/source/src/driver.lo /build/source/src/driver.c
In file included from /build/source/src/driver.c:24:
/build/source/src/driver_rpc.h:9:10: fatal error: rpc/rpc.h: No such file or directory
    9 | #include <rpc/rpc.h>
      |          ^~~~~~~~~~~
compilation terminated.
make: *** [Makefile:177: /build/source/src/driver.lo] Error 1
```

**Comment by cpcloud** [comment]
The issue is related to https://github.com/NixOS/nixpkgs/pull/95089. I tried using the recommended approach of settings `NIX_CFLAGS_COMPILE` + `NIX_LDFLAGS` but that doesn't work because no matter what `libnvidia-container` wants to build and install every target. Even with a patch, it's not clear if `nvidia-docker` needs all of the build outputs of `libnvidia-container` or just the shared libraries.

**Comment by cpcloud** [comment]
Looks like `nvidia-docker` definitely needs the `nvidia-container-cli` binary :(

**Comment by cpcloud** [comment]
Ok, I was able to get a successful build of the SOs and the binary by using a sort-of-gross patch to the `libnvidia-container` top-level `Makefile`.

**Comment by GuillaumeDesforges** [comment]
> The issue is related to #95089. I tried using the recommended approach of settings `NIX_CFLAGS_COMPILE` + `NIX_LDFLAGS` but that doesn't work because no matter what `libnvidia-container` wants to build and install every target. Even with a patch, it's not clear if `nvidia-docker` needs all of the build outputs of `libnvidia-container` or just the shared libraries.

> Looks like `nvidia-docker` definitely needs the `nvidia-container-cli` binary :(

As I understand it, you investigated whether it was possible to skip the phase that required rpcgen or not, correct ?
Thanks for doing that 👍 

> Ok, I was able to get a successful build of `nvidia-docker` by using a sort-of-gross patch to the `libnvidia-container` top-level `Makefile`.

Could you please share your finding ? It could help either propose a fix or find the root cause.

**Comment by cpcloud** [comment]
@GuillaumeDesforges Yeah, I was looking into whether I could build `nvidia-docker` using `libtirpc` as per the recommendation in #95089. You still need the `rpcgen` phase I believe, but as you found adding `rpcsvc-proto` is sufficient to get past that part of the build.

What I found was not very encouraging:

First, I tried depending on `libtirpc`. This doesn't work because the `Makefile` that ships with `libnvidia-container` will build the `shared`, `static`, and `tools` builds. `static` is problematic because the `libtirpc` package doesn't contain static libraries (it's also a different version [1.1.4] than what nixpkgs contains, though I don't know if that's problematic).

I tried patching the `Makefile` enough to get only the `shared` and `tools` targets to build, and then realized that I needed to set the `-DWITH_TIRPC` define compiler flag.

This seems to work, though the patch is quite a hack. I can put up a PR for this if you'd like.

I another approach to avoid patching but it didn't work out and that was to vendor in the right version of libtirpc (using `fetchurl`). That doesn't seem to work because for whatever reason the compiler cannot find the `libtirpc` includes even though they exist in a location that AFAICT was consistent with how the Makefile operates. This of course probably wasn't actually true but by this point I was pretty annoyed and didn't feel like digging any further.
